Cabal/tests/PackageTests/PackageTester.hs

```diff
@@ -1,367 +1,540 @@
-
-doCabalHaddock :: SuiteConfig -> PackageSpec -> [String] -> IO Result
-doCabalHaddock config spec extraArgs = do
-    configResult <- doCabalConfigure config spec
-    if successful configResult
-        then do
-            res <- cabal config spec [] ("haddock" : extraArgs)
-            return $ recordRun res HaddockSuccess configResult
-        else
-            return configResult
-
-unregister :: SuiteConfig -> String -> IO ()
-unregister config libraryName = do
-    res@(_, _, output) <- run Nothing (ghcPkgPath config) []
-                          ["unregister", "--user", libraryName]
-    if "cannot find package" `isInfixOf` output
-        then return ()
-        else requireSuccess res
-
--- | Install this library in the user area
-cabal_install :: SuiteConfig -> PackageSpec -> IO Result
-cabal_install config spec = do
-    buildResult <- doCabalBuild config spec
-    res <- if successful buildResult
-        then do
-            res <- cabal config spec [] ["install"]
-            return $ recordRun res InstallSuccess buildResult
-        else
-            return buildResult
-    record spec res
-    return res
-
-cabal_register :: SuiteConfig -> PackageSpec -> [String] -> IO Result
-cabal_register config spec extraArgs = do
-    res <- doCabalRegister config spec extraArgs
-    record spec res
-    return res
-
-doCabalRegister :: SuiteConfig -> PackageSpec -> [String] -> IO Result
-doCabalRegister config spec extraArgs = do
-    configResult <- doCabalConfigure config spec
-    if successful configResult
-        then do
-            buildResult <- doCabalBuild config spec
-            if successful buildResult
-              then do res <- cabal config spec [] ("register" : extraArgs)
-                      return $ recordRun res RegisterSuccess configResult
-              else return buildResult
-        else
-            return configResult
-
-
-cabal_test :: SuiteConfig -> PackageSpec -> [(String, Maybe String)]
-           -> [String] -> IO Result
-cabal_test config spec envOverrides extraArgs = do
-    res <- cabal config spec envOverrides ("test" : extraArgs)
-    let r = recordRun res TestSuccess nullResult
-    record spec r
-    return r
+------------------------------------------------------------------------
+-- * Compiling setup scripts
+
+compileSetup :: TestM ()
+compileSetup = do
+    (suite, test) <- ask
+    pkg_path <- packageDir
+    liftIO $ rawCompileSetup (suiteVerbosity suite) suite (testEnvironment test) pkg_path
+
+rawCompileSetup :: Verbosity -> SuiteConfig -> [(String, Maybe String)] -> FilePath -> IO ()
+rawCompileSetup verbosity suite e path = do
+    r <- rawRun verbosity (Just path) (ghcPath suite) e $
+        [ "--make"] ++
+        ghcPackageDBParams (packageDBStack suite) ++
+        [ "-hide-all-packages"
+        , "-package base"
+        -- This is better, but we don't necessarily have it
+        -- if we're bootstrapping with old Cabal.
+        -- , "-package-id " ++ CURRENT_PACKAGE_KEY
```

Comment:
You can do that, and that's what we used to do. As before, the problem is that if you do this, you need to configure Cabal twice (once with the old Cabal to actually build it with old Cabal, once with the new Cabal for the test suite to open it with LBI. There's a long comment in PackageTests.hs explaining the situation.)